Add `insert_item` method to ItemList

djdanlib opened the pull request:
Hello, first-time PR here.

With this addition, you can insert an item into an `ItemList` at any index, such as 0 for the top, even including the new index at the bottom that will exist when an item is added. This is a pretty familiar idea for people coming from other GUI frameworks.

Hopefully I've done a decent enough job to consider adding this to the main codebase because I'd love to use this contribution myself. It does compile and work for me on Windows, VS2022, built with scons platform=windows dev_build=yes. I do need some help with the unit testing aspect of this because I don't quite understand how to write an appropriate test for this yet.

Rationale: While working with ItemList, I wanted a method to add/insert an item at any arbitrary index in the list, particularly at the top (index 0), so I created it. It's more efficient (in terms of CPU cycles and lines-of-code) than calling `add_item`, storing its return value, then calling `move_item`.

Code notes: This is short and simple, inspired by `add_item` and `move_item`. I thought about overloading `add_item` for this but landed on creating a separate `insert_item` method instead. Generally the index parameter is first in most languages' insert functions, so that's what I did here - and overloading `add_item` this way didn't feel good. The method doesn't return a value because you either know the index already or will be receiving an error, and as far as I can see, something like `ERR_FAIL_INDEX` isn't available for methods returning a value anyway.

edit: Following a code review, I reworded position to index, since position is commonly used for 2D coordinates. I also removed everything related to an unnecessary insert-at-end modality.

*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/12482

djdanlib committed:
Add insert_item method to item_list
Document new insert_item method in item_list

AThousandShips commented:
Thanks for opening a pull request!

Feature pull requests should be associated to a feature proposal to justify the need for implementing the feature in Godot core. Please open a proposal on the [Godot proposals](https://github.com/godotengine/godot-proposals) repository (and link to this pull request in the proposal's body).

lodetrick commented on doc/classes/ItemList.xml:
```diff
@@ -179,6 +179,18 @@
 		</method>
+		<method name="insert_item">
+			<return type="void" />
+			<param index="0" name="position" type="int" />
```
```suggestion
			<param index="0" name="idx" type="int" />
```
Other methods like `move_item` uses `idx` in their documentation

djdanlib replied:
That's fair. On further reflection, 'position' as used elsewhere usually seems to imply a Vector2.

lodetrick commented on scene/gui/item_list.cpp:
```diff
@@ -71,6 +71,28 @@ int ItemList::add_item(const String &p_item, const Ref<Texture2D> &p_texture, bo
 }
 
+void ItemList::insert_item(int position, const String &p_item, const Ref<Texture2D> &p_texture, bool p_selectable) {
+	if (position == items.size()) {
+		add_item(p_item, p_texture, p_selectable);
+		return;
+	}
+	ERR_FAIL_INDEX(position, items.size());
```
```suggestion
	ERR_FAIL_INDEX(position, items.size() + 1);
```
`Vector::insert` supports appending if the index is equal to the length

djdanlib replied:
Great catch!

djdanlib commented:
Thanks @lodetrick for reviewing my code and teaching+reminding me a few things!

KoBeWi reviewed (approved):
The implementation looks ok.

Not sure how useful is the new method, but the proposal does have some support and makes a good point about examples in other GUI libs.

Though the ugly code mentioned in the proposal can be replaced with a helper method, like `Utils.insert_item(list, idx, text)` (where Utils is a class or autoload singleton) etc.

The commits need to be squashed.

Mickeon commented:
The existence of this PR fascinates me, because I somewhat assumed this method did not exist for consistency with Node's own API, where adding and moving are separate methods, as well.

djdanlib commented:
Should I also do an `insert_icon_item` to match `add_icon_item` for completeness?

Respect to the real Godot devs here, I feel like such an imposter!

I consider a list box in any UI toolkit to be an interface to a collection. Some toolkits expose the underlying collection. When they don't, I'd typically want access to basic collection methods.

This certainly could be done in a way that keeps the context visually clean by refactoring those two lines out to a helper class. That's what I was about to do before going this route. You'd lose even more efficiency than just using those two lines though. Now you have a new class involved, which feels cluttery, like work I shouldn't be doing. Then you've spent time to write a class, added performance overhead, and added mental overhead. I'm all for reducing complexity. Let's put more wind in a dev's sails, which drives adoption and potential contribution, right?

You're right, there isn't exactly a similar thing in `Node`. That's a little surprising if you assume the point of view of someone who just wants to sit down and make stuff. Maybe it's just because nobody's done it yet? `Node` has an indexed `get_child` and `move_child`, and if you tracked/iterated children (which are both annoying things to have to do) you could use `add_sibling`, but there isn't something that maps directly to the concept of an indexed insert. (There also isn't a corresponding indexed remove child operation.) It could be done, I suppose, but to me that looks difficult. Node uses a HashMap for its children under the hood. I'm not brave enough to attempt the level of change it looks like an indexed insert would entail - seems to be a fairly lengthy stack of functions would have to be modified. And I might just be rambling off-topic.

djdanlib commented:
I just realized that I was meant to do the squashing. Sorry, I just took care of that, it should be all set now.

AThousandShips commented on doc/classes/ItemList.xml:
```diff
@@ -179,6 +179,18 @@
+			<param index="2" name="icon" type="Texture2D" default="null" />
+			<param index="3" name="selectable" type="bool" default="true" />
+			<description>
+				Inserts an item at the specified index in the item list with the specified text.
```
```suggestion
				Inserts an item with the specified text at the specified index in the item list.
```
Otherwise it seems like the item list has the specified text

djdanlib committed:
Minor wording update for doc/classes/ItemList.xml

Thanks to AThousandShips for the suggestion!

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>
